resolwe_bio/processes/expression/featureCounts.py

```diff
@@ -0,0 +1,888 @@
+            args.append("-L")
+
+        if inputs.miscellaneous.report_reads:
+            args.append("-R CORE")
+
+        if inputs.miscellaneous.verbose:
+            args.append("--verbose")
+
+        # for paired-end samples, extend the args with paired-end input options
+        if paired_end and inputs.paired_end.is_paired_end:
+            args.append("-p")
+
+        if paired_end and inputs.paired_end.require_both_ends_mapped:
+            args.append("-B")
+
+        if paired_end and inputs.paired_end.check_frag_length:
+            args.append("-P")
+
+        if paired_end and inputs.paired_end.check_frag_length:
+            args.extend(["-d", inputs.paired_end.min_frag_length])
+
+        if paired_end and inputs.paired_end.check_frag_length:
+            args.extend(["-D", inputs.paired_end.max_frag_length])
+
+        if paired_end and inputs.paired_end.do_not_count_chimeric_fragments:
+            args.append("-C")
+
+        if paired_end and inputs.paired_end.do_not_sort:
+            args.append("--do_not_sort")
+
+        # Run featureCounts
+        return_code, _, _ = Cmd["featureCounts"][args][bam_file] & TEE(retcode=None)
+        if return_code:
+            self.error("Error while running featureCounts.")
+
+        # parse featureCounts output
+        get_gene_counts("featureCounts_rc.txt", "rc.txt", bam_file)
+        get_gene_lenghts("featureCounts_rc.txt", "gene_lengths.txt")
+
+        # Normalize counts
+        rnanorm_args = [
+            "rc.txt",
+            "--gene-lengths",
+            "gene_lengths.txt",
+            "--tpm-output",
+            "tpm.txt",
+            "--cpm-output",
+            "cpm.txt",
+        ]
+        return_code, _, _ = Cmd["rnanorm"][rnanorm_args] & TEE(retcode=None)
+        if return_code:
+            self.error("Error while normalizing counts using rnanorm.")
+
+        # prepare the expression set outputs
+        feature_ids = pd.read_csv(
+            "rc.txt", sep="\t", index_col="FEATURE_ID"
+        ).index.to_list()
+
+        feature_filters = {
+            "source": inputs.annotation.output.source,
+            "species": inputs.aligned_reads.output.species,
+            "feature_id__in": feature_ids,
+        }
+
+        feature_ids_to_names = {
+            f.feature_id: f.name for f in self.feature.filter(**feature_filters)
+        }
+
+        prepare_expression_set(
+            rc="rc.txt",
+            tpm="tpm.txt",
+            cpm="cpm.txt",
+            feature_dict=feature_ids_to_names,
+            outfile_name=f"{name}_expressions",
+        )
+
+        # rename and compress the expression files
+        rename_columns_and_compress("rc.txt", f"{name}_rc.tab.gz")
+        rename_columns_and_compress("tpm.txt", f"{name}_tpm.tab.gz")
+        rename_columns_and_compress("cpm.txt", f"{name}_cpm.tab.gz")
+
+        # Compress the featureCounts output file
+        (
+            Cmd["gzip"]["-c", "featureCounts_rc.txt"]
```

Comment:
Technically this could be rewritten into a more pythonic way, without resorting to `Cmd`.

```python
import gzip
import shutil

with open(filename="featureCounts_rc.txt", mode = "rb") as f_in:
    with gzip.open(filename=f"{name}_featureCounts_rc.txt.gz", mode="wb") as f_out:
        shutil.copyfileobj(f_in, f_out)
```

Reply:
Done